src/main/java/org/simplify4u/plugins/ArtifactResolver.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright 2017 Slawomir Jaranowski
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.simplify4u.plugins;
+
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.artifact.repository.ArtifactRepository;
+import org.apache.maven.artifact.resolver.ArtifactResolutionRequest;
+import org.apache.maven.artifact.resolver.ArtifactResolutionResult;
+import org.apache.maven.model.Dependency;
+import org.apache.maven.model.Plugin;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.logging.Log;
+import org.apache.maven.project.MavenProject;
+import org.apache.maven.repository.RepositorySystem;
+import org.simplify4u.plugins.skipfilters.SkipFilter;
+
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.StreamSupport.stream;
+
+/**
+ * Artifact resolver for project dependencies, build plug-ins, and build plug-in dependencies.
+ */
+final class ArtifactResolver {
+
+    private final Log log;
+
+    private final RepositorySystem repositorySystem;
+
+    private final ArtifactRepository localRepository;
+
+    private final List<ArtifactRepository> remoteRepositories;
+
+    ArtifactResolver(final Log log, final RepositorySystem repositorySystem,
```

Comment:
I not prefer use final for method arguments, it obscure code. Of course rewriting parameters is not good - bot this it will be taken by sonar.